python_modules/dagster/dagster/core/asset_defs/asset_group.py

```diff
@@ -429,16 +422,33 @@ def from_modules(
         Returns:
             AssetGroup: An asset group with all the assets defined in the given modules.
         """
-        assets: Set[AssetsDefinition] = set()
-        source_assets: Set[SourceAsset] = set()
+        asset_ids: Set[int] = set()
+        asset_keys: Set[AssetKey] = set()
+        source_assets: List[SourceAsset] = []
+        assets: List[AssetsDefinition] = []
         for module in modules:
             module_assets, module_source_assets = _find_assets_in_module(module)
-            assets.update(module_assets)
-            source_assets.update(module_source_assets)
+            all_assets: List[Union[AssetsDefinition, SourceAsset]] = [
+                *module_assets,
+                *module_source_assets,
+            ]
+            for asset in all_assets:
+                if not id(asset) in asset_ids:
+                    asset_ids.add(id(asset))
+                    keys = asset.asset_keys if isinstance(asset, AssetsDefinition) else [asset.key]
+                    for key in keys:
+                        if key in asset_keys:
+                            raise DagsterInvalidDefinitionError(
+                                f"Asset key {key} is defined multiple times in module {module.__name__}"
+                            )
+                    if isinstance(asset, SourceAsset):
```

Comment:
A slightly less verbose way of doing this, that also would allow us to avoid doing another `isinstance` check, which always feels a little dirty, would be:
```
source_assets.extend(module_source_assets)
assets.extend(source_assets)
```
a level of indentation back.

Reply:
Hmm but this doesn't handle deduping right? We only want to add to the the list if the object's id hasn't already been seen.

Agree that double-isinstance checking is undesirable but IIRC it was between that and repeating (or factoring out) the error.

Reply:
Ah yes, got it. If we are going to smush them together, then might make sense to just have `_find_assets_in_module` produce a generator of `Union[SourceAsset, AssetsDefinition]`? That would make it simpler and avoid creating intermediate data structures.